File: go/cli/output.go
package cli

import (
	"fmt"
	"io"
	"os"
)

// writeOut writes content to the file at outPath, or to w (the command's stdout
// writer) when outPath is empty. Shared by `compile` (Go source) and `encode`
// (JSON text or binary bytes). No trailing newline is added beyond what content
// already carries.
func writeOut(w io.Writer, outPath string, content []byte) *cliError {
	if outPath == "" {
		if _, err := w.Write(content); err != nil {
			return ioErr("could not write to stdout: %v", err)
		}
		return nil
	}
	if err := os.WriteFile(outPath, content, 0o644); err != nil {
		return ioErr("could not write %s: %v", outPath, err)
	}
	return nil
}

// printLine writes a single line plus a newline to w, surfacing a write failure
// as an I/O error (exit 3) rather than swallowing it.
func printLine(w io.Writer, line string) *cliError {
	if _, err := fmt.Fprintln(w, line); err != nil {
		return ioErr("could not write to stdout: %v", err)
	}
	return nil
}

File: python/cli/ball_cli/output.py
"""Shared output helper: write to ``-o <file>`` or the command's stdout."""

from __future__ import annotations

from pathlib import Path
from typing import TextIO

from .errors import io_error


def write_output(out_path: str | None, content: str, stdout: TextIO) -> None:
    """Write ``content`` to ``out_path``, or to ``stdout`` when it is falsy.

    Shared by ``compile`` (Python source) and ``encode`` (JSON text). A write
    failure to ``--output`` is an I/O error (exit 3)."""
    if out_path:
        try:
            Path(out_path).write_text(content, encoding="utf-8")
        except OSError as ex:
            raise io_error(f"could not write {out_path}: {ex}") from ex
    else:
        stdout.write(content)
